src/main/java/com/netflix/imflibrary/st2067_2/ApplicationCompositionFactory.java

```diff
@@ -101,22 +101,30 @@ public static ApplicationComposition getApplicationComposition(ResourceByteRange
 
         try {
             IMFCompositionPlaylistType imfCompositionPlaylistType = IMFCompositionPlaylistType.getCompositionPlayListType(resourceByteRangeProvider, imfErrorLogger);
-            String applicationIdentification = imfCompositionPlaylistType.getApplicationIdentification();
-            ApplicationCompositionType applicationCompositionType = ApplicationCompositionType.fromApplicationID(applicationIdentification);
-
-            if(applicationCompositionType == null) {
+            if (imfCompositionPlaylistType.getApplicationIdentificationSet().size() == 0) {
                 clazz = Application2ExtendedComposition.class;
                 imfErrorLogger.addError(IMFErrorLogger.IMFErrors.ErrorCodes.IMF_CPL_ERROR, IMFErrorLogger.IMFErrors.ErrorLevels.NON_FATAL,
-                    String.format("Unsupported/Missing ApplicationIdentification %s in CPL", applicationIdentification));
-            }
-            else
-            {
-                clazz = applicationCompositionType.getClazz();
+                        String.format("Missing ApplicationIdentification in CPL"));
+                Constructor<?> constructor = clazz.getConstructor(IMFCompositionPlaylistType.class, Set.class);
+                composition = (ApplicationComposition) constructor.newInstance(imfCompositionPlaylistType, homogeneitySelectionSet);
+                imfErrorLogger.addAllErrors(composition.getErrors());
+            } else {
+                for (String applicationIdentification : imfCompositionPlaylistType.getApplicationIdentificationSet()) {
+                    ApplicationCompositionType applicationCompositionType = ApplicationCompositionType.fromApplicationID(applicationIdentification);
+
+                    if (applicationCompositionType == null) {
+                        clazz = Application2ExtendedComposition.class;
+                        imfErrorLogger.addError(IMFErrorLogger.IMFErrors.ErrorCodes.IMF_CPL_ERROR, IMFErrorLogger.IMFErrors.ErrorLevels.NON_FATAL,
+                                String.format("Unsupported/Missing ApplicationIdentification %s in CPL", applicationIdentification));
+                    } else {
+                        clazz = applicationCompositionType.getClazz();
+                    }
+
+                    Constructor<?> constructor = clazz.getConstructor(IMFCompositionPlaylistType.class, Set.class);
+                    composition = (ApplicationComposition) constructor.newInstance(imfCompositionPlaylistType, homogeneitySelectionSet);
```

Comment:
Isn't `composition` being overwritten each time if there is more than one _ApplicationIdentification_? That also mean that the order that AppId are provided matters, since only the last composition is returned.

I'm not familiar with the use of multiple ApplicationIdentifications, but is the intent to still return a single composition that conforms to ALL of them? Or a map of compositions, one for each AppId?

Reply:
You're right that the composition would be overwritten and therefore that the returned object would be dependent on the order of applications declared. In the current approach, the getter actually does a validation of the CPL, so the idea is that it does a validation against all declared applications, even if it returns only the last one. This is not ideal, but in practice, all CPLs we have seen so far declare a single applications. So the result will be identical. In the future, we could introduce a different getter to force which application is returned.